### PR TITLE
Simplify parametrizations.SpectralNorm and improve its initialization

### PR DESCRIPTION
**Pull request by lezcano**
Implements a number of changes discussed with @soulitzer offline.
In particular:
- Initialise `u`, `v` in `__init__` rather than in `_update_vectors`
- Initialise `u`, `v` to some reasonable vectors by doing 15 power iterations at the start
- Simplify the code of `_reshape_weight_to_matrix` (and make it faster) by using `flatten`

### CODE REVIEW

**Review by soulitzer — COMMENTED** [review]
I would mention in that docs that we now iterate a couple times on initialization, but looks good otherwise! Thanks for the change.

File: torch/nn/utils/parametrizations.py [hunk 1]
```diff
@@ -20,23 +20,25 @@ def __init__(
...
-        self._update_vectors(weight_mat)
+        h, w = weight_mat.size()
+
+        self.register_buffer('u',
+            F.normalize(weight_mat.new_empty(h).normal_(0, 1), dim=0, eps=self.eps))
+        self.register_buffer('v',
+            F.normalize(weight_mat.new_empty(w).normal_(0, 1), dim=0, eps=self.eps))
+
+        # Start with u, v initialized to some reasonable values
+        self._update_vectors(weight_mat, 15)
```
Comment (soulitzer) [inline torch/nn/utils/parametrizations.py:32]: Maybe link to the paper where its specified, or small derivation/explanation on why its reasonable.

File: torch/nn/utils/parametrizations.py [hunk 2]
```diff
@@ -67,17 +69,12 @@ def _update_vectors(self, weight_mat: torch.Tensor) -> None:
...
 
-        for _ in range(self.n_power_iterations):
+        for _ in range(n_power_iterations):
             # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
             # are the first left and right singular vectors.
             # This power iteration produces approximations of `u` and `v`.
-            self.u = F.normalize(torch.mv(weight_mat, self.v),
+            self.u = F.normalize(torch.mv(weight_mat, self.v),      # type: ignore[has-type]
                                  dim=0, eps=self.eps, out=self.u)   # type: ignore[has-type]
             self.v = F.normalize(torch.mv(weight_mat.t(), self.u),  # type: ignore[has-type]
```
Comment (soulitzer) [inline torch/nn/utils/parametrizations.py:79]: Were you going to make this change as well, why did you decide against it?

Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:79]: You mean the one about the multiplication order to get a better memory layout?

I realised that `mv`, at least in [LAPACK](https://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_gaeda3cbd99c8fb834a60a6412878226e1.html), does have a flag that tells the algorithm that the matrix is transposed so, I would assume that PyTorch handles this correctly and makes this optimisation internally. That being said, I have not looked into that just yet, I'll do that tomorrow.

Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:79]: So it does not make use of this, but I plan to implement this optimisation soon https://github.com/pytorch/pytorch/issues/59785

**Comment by vadimkantorov** [comment]
Also there is new support for parametrizations: https://github.com/pytorch/pytorch/issues/7313

+ it may be good to factor out power iteration to a fully-fledged method? (at least in utils for now?) https://github.com/pytorch/pytorch/issues/36314 https://github.com/pytorch/pytorch/issues/12760

**Comment by Lezcano** [comment]
This could be interesting. Could you provide speed / accuracy benchmarks for the proposed change? It is not clear a priori what's the overhead of lowrank solvers have against one or two power method iterations, even less in this setting where the eigen-gap might be very small. Same happens with the repercussion that this change would have on GAN models.

**Comment by vadimkantorov** [comment]
For lobpcg - need to be tested for sure, but I think factoring out this power iteration method to a more visible (even if semi-experimental) torch.nn.utils function is a good idea I think - then would be easier to swap it out and check convergence / precision etc

**Comment by Lezcano** [comment]
@albanD @soulitzer This should be ready for another round of reviews.

File: test/test_nn.py [hunk 3]
```diff
@@ -3809,10 +3810,18 @@ def get_modules():
 
                 # TEST TRAINING BEHAVIOR
 
-                # run forward again and assert that u and v are updated
+                # We perform GD first to modify the initial matrix
+                wrapped_m(input).sum().backward()
+                with torch.no_grad():
+                    for p in wrapped_m.parameters():
+                        if p.requires_grad:
+                            p.add_(- p.grad, alpha=0.01)
```
Comment (soulitzer) [inline test/test_nn.py:3818]: nit: extra space here?

File: torch/nn/utils/parametrizations.py [hunk 4]
```diff
@@ -86,11 +97,15 @@ def _update_vectors(self, weight_mat: torch.Tensor) -> None:
         self.v = self.v.clone(memory_format=torch.contiguous_format)
 
     def forward(self, weight: torch.Tensor) -> torch.Tensor:
-        weight_mat = self._reshape_weight_to_matrix(weight)
-        if self.training:
-            self._update_vectors(weight_mat)
-        sigma = torch.dot(self.u, torch.mv(weight_mat, self.v))
-        return weight / sigma
+        if weight.ndim == 1:
```
Comment (soulitzer) [inline torch/nn/utils/parametrizations.py:100]: Is there a use case for the `ndim = 1` case? Wondering if its more likely that someone registering spectral_norm for vector is a user error - in that case, might be better to just have an explicit error instead of silently doing something unexpected.

Comment (soulitzer) [reply torch/nn/utils/parametrizations.py:100]: Alternatively: just add a note in the doc mentioning behavior on 1D

File: test/test_nn.py [hunk 5]
```diff
@@ -3809,10 +3810,18 @@ def get_modules():
...
+
                 out = wrapped_m(input)
-                self.assertNotEqual(u0, spectral_norm_m.u)
-                self.assertNotEqual(v0, spectral_norm_m.v)
+                if requires_grad:
+                    # run forward again and assert that u and v are updated
+                    self.assertNotEqual(u0, spectral_norm_m.u)
```
Comment (soulitzer) [inline test/test_nn.py:3823]: I'd probably agree with @albanD here, with having `u` and `v` be `_u` and `_v` instead. Even though `_SpectralNorm` is private, `m.parametrizations.weight[0]` seems to be a public api no?

File: test/test_nn.py (hunk 3 above)
Comment (Lezcano) [reply test/test_nn.py:3818]: I actually went on an changed this code with calls to `SGD` as suggested by @albanD in https://github.com/pytorch/pytorch/pull/58488#discussion_r647458661

File: torch/nn/utils/parametrizations.py (hunk 4 above)
Comment (Lezcano) [reply torch/nn/utils/parametrizations.py:100]: I'll add the note

**Comment by Lezcano** [comment]
@soulitzer This is ready for another round of reviews.

I was looking into using `torch.linalg.multi_dot` to compute the multiplication, but in the end I decided against it, as it uses `mm`, while here we can use the faster `mv`. I was going to then implement a simple `multi_dot` for this case myself (i.e. choose the order of the multiplications) but then I realised that the improvement would be minimal, so I chose to leave things as they are, as the code is simpler this way.

**Review by soulitzer — APPROVED** [review]
LGTM

edit: You might want to rebase to fix the merge conflicts

**Comment by Lezcano** [comment]
@soulitzer rebased!
